controllers/asosecret_controller.go

```diff
@@ -278,14 +278,19 @@ func (asos *ASOSecretReconciler) createSecretFromClusterIdentity(ctx context.Con
 	newASOSecret.Data["AZURE_TENANT_ID"] = []byte(identity.Spec.TenantID)
 	newASOSecret.Data["AZURE_CLIENT_ID"] = []byte(identity.Spec.ClientID)
 
+	// If the identity type is WorkloadIdentity or UserAssignedMSI, then we don't need to fetch the secret so return early
+	if identity.Spec.Type == infrav1.WorkloadIdentity || identity.Spec.Type == infrav1.UserAssignedMSI {
+		return newASOSecret, nil
+	}
+
 	// Fetch identity secret, if it exists
 	key = types.NamespacedName{
 		Namespace: identity.Spec.ClientSecret.Namespace,
 		Name:      identity.Spec.ClientSecret.Name,
 	}
 	identitySecret := &corev1.Secret{}
 	err := asos.Get(ctx, key, identitySecret)
-	if err != nil && !apierrors.IsNotFound(err) {
```

Comment:
Are not found errors handled differently now that the secret is being returned early, or is this solving a different problem?

Reply:
This is solving a different problem. If secret is not found, we need to return here too in order to avoid any issues few lines above were we are referencing identity secret data that won't be available if the secret doesn't exist, e.g. `identitySecret.Data`